users/config: Structured option documentation. (#710)
syncthing/docs: users/config.rst

    It is customary that the local device ID is included in all folders.
    Syncthing will currently add this automatically if it is not present in
    the configuration file.

    The ``encryptionPassword`` sub-element contains the secret needed to decrypt
    this folder's data on the remote device.  If left empty, the data is plainly
    accessible (but still protected by the transport encryption).  The mechanism
    and how to set it up is described in a :doc:`separate chapter <untrusted>`.

.. option:: folder.minDiskFree

    The minimum required free space that should be available on the disk this
    folder resides.  The folder will be stopped when the value drops below the
    threshold.  The element content is interpreted according to the given
    ``unit`` attribute.  Accepted ``unit`` values are ``%`` (percent of the disk
    / volume size), ``kB``, ``MB``, ``GB`` and ``TB``.  Set to zero to disable.

.. option:: folder.versioning

    Specifies a versioning configuration.

    .. seealso::
	:doc:`versioning`

.. option:: folder.copiers
	    folder.hashers

    The number of copier and hasher routines to use, or ``0`` for the

>     system determined optimums. These are low level performance options for

Reviewer: ```suggestion
    system determined optimums. These are low-level performance options for
```

    advanced users only; do not change unless requested to or you've actually
    read and understood the code yourself. :)

.. option:: folder.pullerMaxPendingKiB

    Controls when we stop sending requests to other devices once we’ve got this
    much unserved requests.  The number of pullers is automatically adjusted
    based on this desired amount of outstanding request data.

.. option:: folder.order

>     The order in which needed files should be pulled from the cluster.

Reviewer: ```suggestion
    The order in which needed files should be pulled from remote devices.
```

    The possibles values are:

    ``random`` (default)
        Pull files in random order. This optimizes for balancing resources among

>         the devices in a cluster.

Reviewer: Same as above, but I'm not sure what to change "cluster" to here. Maybe "among the connected devices" or "among the devices sharing the same folder"?

Author: What's wrong with the word "cluster"?  It's technically correct, though not a technical term: "a collection of things somehow tied together".  If there's no obvious better term, let's just leave it at that.

Note that we're looking at the detailed documentation for an XML file format here.  As such we don't necessarily need to aim for Average Joe to understand everything.

Reviewer: I treat "cluster" similarly to "node" and "repository". All are present in the English dictionary, but in reality they're not used by the general population (when it comes to computers and such). As for Syncthing, "nodes" were replaced by "devices", "repositories" by "folders" a few years ago. "Cluster" was left out in the process.

I just think it's beneficial to use simpler terms unless there is no other option. While the average Joe is unlikely to come and explore the docs on his own, he is very often sent here from the forum, so the easier to understand the explanations are, the better.

Author: Well, "node" is closely related to graphs and "repository" is technical, both rather uncommon words.  

> "Cluster" was left out in the process.

And I think it's still ok to use.

Reviewer: "Cluster", at least regarding computing, is on the same level as the other two, really. I mean, there's a reason why cloud/sync solutions aimed at the general public (e.g. Dropbox, Google Drive, OneDrive, etc.) don't use such terminology in their documentation, unless it's from the get-go very clearly directed strictly at developers and/or power users.

I'm not saying this needs to be addressed in this commit, but the question remains whether the documentation (and wording in general) should be accessible to all users of Syncthing, including non-technical people, or only targeted at techies and such. I'd also keep in mind that we're in the general "usage" section of the Docs here, not "development" or even "advanced".

Author: 1. Dropbox etc. use a client-server model, where "cluster" is not as meaningful as for a decentralized system.
2. This page explains how to interpret entries in an XML file.  Possibly will get linked to from the "advanced" configuration panel.  It's directly strictly at developers and power users, otherwise we should definitely get rid of the large XML quotes :wink:

Glad to hear you don't insist on this discussion being finished within this PR.

Reviewer: > Dropbox etc. use a client-server model

Dropbox is a little more complex than that, e.g. it also supports direct sync between LAN devices without downloading files from the central server, but let's not go into such details here.

> This page explains how to interpret entries in an XML file.  Possibly will get linked to from the "advanced" configuration panel.

Yeah, this is a different problem, I'd say mainly about a discrepancy between the Web GUI and the Docs. The two are structured differently, and what's clearly divided in the GUI between "basic" and "advanced" configuration, in the Docs is all simply presented at the same page. As such, it becomes the source of information for both power users and beginners. I don't think there's an easy solution to this problem, but I personally just always prefer to stick to using language that is easy and understandable by everyone (which includes beginners, of course, but also non-native English speakers, especially since the Docs are not available in other languages yet).

Author: Actually there are a few help links from the regular "Settings" dialog which lead directly to the detailed sub-pages here (e. g. GUI listen address, Automatic upgrades).  But others link to the config description, which could be improved.  There simply isn't a "novice-compatible" help topic for everything.  Feel free to write up more stuff in more easily understood language and submit it.  The distinction itself makes sense to me, because no regular user wants to scroll through XML code.

.. option:: folder.minDiskFree

    The minimum required free space that should be available on the disk this
    folder resides.  The folder will be stopped when the value drops below the
    threshold.  The element content is interpreted according to the given
    ``unit`` attribute.  Accepted ``unit`` values are ``%`` (percent of the disk
    / volume size), ``kB``, ``MB``, ``GB`` and ``TB``.  Set to zero to disable.

.. option:: folder.versioning

    Specifies a versioning configuration.

    .. seealso::
	:doc:`versioning`

.. option:: folder.copiers
	    folder.hashers

    The number of copier and hasher routines to use, or ``0`` for the
    system determined optimums. These are low-level performance options for
    advanced users only; do not change unless requested to or you've actually
    read and understood the code yourself. :)

.. option:: folder.pullerMaxPendingKiB

    Controls when we stop sending requests to other devices once we’ve got this
    much unserved requests.  The number of pullers is automatically adjusted
    based on this desired amount of outstanding request data.

.. option:: folder.order

>     The order in which needed files should be pulled from the cluster.

Reviewer: ```suggestion
    The order in which needed files should be pulled from the cluster. 
    It has no effect when the folder type is "send only".
```

Reviewer: Just one quick addition I'd like to suggest (actually taken from https://github.com/syncthing/docs/pull/532).

        sequences at random. Provides acceptable data distribution and minimal spinning disk strain.

    ``random``
        The blocks of a file are downloaded in a random order. Provides great data distribution, but very taxing on
        spinning disk drives.

    ``inOrder``
        The blocks of a file are downloaded sequentially, from start to finish. Spinning disk drive friendly, but provides
        no improvements to data distribution.

.. option:: folder.copyRangeMethod

    Provides a choice of method for copying data between files.  This can be
    used to optimise copies on network filesystems, improve speed of large
    copies or clone the data using copy-on-write functionality if the underlying
    filesystem supports it.  The mechanism is described in a :doc:`separate
    chapter </advanced/folder-copyrangemethod>`.

.. option:: folder.caseSensitiveFS

    Affects performance by disabling the extra safety checks for case
    insensitive filesystems.  The mechanism and how to set it up is described in
    a :doc:`separate chapter </advanced/folder-caseSensitiveFS>`.

.. option:: folder.junctionsAsDirs

>     NTFS directory junctions can be treated as ordinary directories, if this is

Reviewer: ```suggestion
    NTFS directory junctions are treated as ordinary directories, if this is
```

            <address>dynamic</address>
        </device>

.. option:: device.paused

    True if synchronization with this devices is (temporarily) suspended.

.. option:: device.allowedNetwork

    If given, this restricts connections to this device to only this network.
    The mechanism is described in detail in a :doc:`separate chapter
    </advanced/device-allowednetworks>`).

.. option:: device.autoAcceptFolders

    If ``true``, folders shared from this remote device are automatically added
    and synced locally under the :opt:`default path <defaults.folder>`.  For the
    folder name, Syncthing tries to use the label from the remote device, and if
    the same label already exists, it then tries to use the folder's ID.  If

>     that exists as well, it just logs an error.  A local folder already added

Reviewer: I think it actually asks to add the folder manually in such a case? Does it still log an error? It's been a while since I tested these, but I don't remember seeing any actual errors. I may be wrong, of course, but if possible, please re-check this one.

Author: You're right.  I just checked the code again and as far as I can tell it just gets the normal treatment.  The log message is just at INFO level.

.. option:: device.maxRecvKbps

    Maximum receive rate to use for this device. Unit is kibibytes/second,
    despite the config name looking like kilobits/second.

.. option:: device.ignoredFolder

    Contains the ID of the folder that should be ignored. This folder will
    always be skipped when advertised from the containing remote device,
    i.e. this will be logged, but there will be no dialog shown in the web GUI.

.. option:: device.maxRequestKiB

    Maximum amount of data to have outstanding in requests towards this device.
    Unit is kibibytes.

.. option:: device.remoteGUIPort

    If set to a positive integer, the GUI will display an HTTP link to the IP
    address which is currently used for synchronization.  Only the TCP port is
    exchanged for the value specified here.  Note that any port forwarding or
    firewall settings need to be done manually and the link will probably not
    work for link-local IPv6 addresses because of modern browser limitations.

.. option:: device.untrusted

    This boolean value marks a particular device as untrusted, which disallows
    ever sharing any unencrypted data with it.  Every folder shared with that

>     device then needs an encryption password set.  Refer to the detailed

Reviewer: I think you can also share receive encrypted folders with such devices, can't you? Those don't need to have a password set on the device that shares them.

Author: I don't know. I followed what was described on the [dedicated page](https://docs.syncthing.net/users/untrusted.html#gui):

> If you never want to sync any unencrypted data to a particular device, you can mark it as untrusted in the advanced tab of the device edit dialog. This will prevent a connection to that device if you forget to set a password on any folder shared with it.

Reviewer 2: The intention is definitely that you can share an already encrypted folder with an untrusted device. I haven't thought about that when implementing though, so not sure it works. I'd consider it a bug though if it didn't.

Reviewer: I've just tested this. I can confirm you can share an already encrypted folder with an untrusted device. The only caveat is that the folder type must be manually adjusted to "receive encrypted" on the untrusted side, but other than that, everything seems to be working as expected.

Author: So, any suggestion what changes are needed in the text?

Author: I committed a possible solution.  It also documents the folder.type = `receiveencrypted`, which was previously missed.

    disable this behavior, for example to control process priority yourself
    as part of launching Syncthing, set this option to ``false``.

.. option:: options.maxFolderConcurrency

    This option controls how many folders may concurrently be in I/O-intensive

>     operations such as syncing or scanning.  The mechanism is described in

Reviewer: Just a side note, as I know this text comes from the already existing subpage, but in most cases syncing is not I/O intensive, especially outside of a local LAN, where bandwidth is almost always the limit, not the I/O. Even on a local LAN, I have personally not seen a case of I/O being the problem. It has always been the connection itself.

Only a digression; nothing to do here.

    detail in a :doc:`separate chapter </advanced/option-max-concurrency>`.

.. option:: options.crashReportingURL

    Server URL where :doc:`automatic crash reports <crashrep>` will be sent if
    enabled.

.. option:: options.crashReportingEnabled

    Switch to opt out from the :doc:`automatic crash reporting <crashrep>`

>     feature to avoid Syncthing "phoning home" on serious troubles.  Defaults to

Reviewer: I'd suggest to make this one explicit, i.e. replace "phoning home" with "sending panic logs" (with a `ref` to https://docs.syncthing.net/users/crashrep.html).

Author: It already is a ref to that page.  "Phoning home" is an established term AFAICT, which should trigger anybody who's sensitive to such behavior in a piece of software.  That was exactly my goal.

Reviewer: "Phoning home" is usually used with a negative connotation though. I don't think you'll find any other piece of software calling its own data or crash collection with this particular term.

Reviewer: Yeah, there will be exceptions. Still, when people hear about "phoning home", it's usually closer to something like https://techsupportalert.com/how-to-stop-windows-10-from-phoning-home or https://news.softpedia.com/news/samsung-phones-said-to-come-with-chinese-spyware-phoning-home-528768.shtml, which implicitly indicates nefarious activities, so I'd still suggest to use a more "neutral" expression here.

Reviewer 2: I agree with Tomasz here: We want people to give a choise, but we also want people to send crash reports. Just the neutral statement he proposed (to prevent Syncthing from sending crash reports) is more appropriate and even has more meaning (phoning home can mean basically anything).

Author: Done.

    ``true``, to help the developers troubleshoot.

.. option:: options.databaseTuning

    Controls how Syncthing uses the backend key-value database that stores the
    index data and other persistent data it needs.  The available options and
    implications are explained in a :doc:`separate chapter
    </advanced/option-database-tuning>`.

.. option:: options.maxConcurrentIncomingRequestKiB

    This limits how many bytes we have "in the air" in the form of response data

>     being read and processed.

Reviewer: I think this one needs to be explained more, in more "human language", if possible. Right now, I can't understand it, let alone any even less technical person.

Author: Me neither.  I'm living off what I could find as my understanding of Syncthing's code is also limited.  Whoever has time / energy to improve it, please do so.  But let's keep that in a separate PR.  My goal was to have completeness and not introduce additional TODO boxes.

.. option:: options.announceLANAddresses

    Enable (the default) or disable announcing private (RFC1918) LAN IP

>     addresses to global discovery.

Reviewer: Maybe add a link to https://datatracker.ietf.org/doc/html/rfc1918?

Author: Honestly, I'd avoid it.  Who knows when it will break and linking to it raises the expectations regarding how closely the standard was followed in Syncthing's code.  Which is probably rather "pragmatic" :wink:

.. option:: options.sendFullIndexOnUpgrade

>     Controls whether all index data is resent when an upgrade is detected,

Reviewer: ```suggestion
    Controls whether all index data is resent when an upgrade has happened,
```

Author: Is that an upgrade of the local Syncthing executable?  If so, your suggestion makes sense.  Though if it also includes upgrades remote devices, then "detect" is probably appropriate.

Reviewer: It is the local executable. I had this option enabled for a few months, and I also remember the same behaviour when it always took place after upgrading Syncthing.

Technically, it doesn't even have to be an "upgrade". Simply replacing the executable with a different one (e.g. a custom-built or an older version) will trigger the index exchange.

    equivalent to starting Syncthing with :option:`--reset-deltas`.  This used

>     to be the default behavior in previous versions, but is mainly useful as a

Reviewer: ```suggestion
    to be the default behavior in older versions, but is mainly useful as a
```
It's been quite a while, so I'd suggest "older".

    troubleshooting step and causes high database churn. The default is now
    ``false``.

.. option:: options.featureFlag

    Feature flags are simple strings that, when added to the configuration, may

>     unleash unfished or still-in-development features to allow early user

Reviewer: ```suggestion
    unleash unfinished or still-in-development features to allow early user
```

Author: Thanks, lol!

    testing.  Any supported value will be separately announced with the feature,
    so that regular users do not enable it by accident.

.. option:: options.connectionLimitEnough

    The number of connections at which we stop trying to connect to more
    devices, zero meaning no limit.  Does not affect incoming connections.  The
    mechanism is described in detail in a :doc:`separate chapter
    </advanced/option-connection-limits>`.

.. option:: options.connectionLimitMax

    The maximum number of connections which we will allow in total, zero meaning
    no limit.  Affects incoming connections and prevents attempting outgoing
    connections.  The mechanism is described in detail in a :doc:`separate
    chapter </advanced/option-connection-limits>`.

.. option:: options.insecureAllowOldTLSVersions

>     Only for compatibility with old devices, as detailed in

Reviewer: ```suggestion
    Only for compatibility with old versions of Syncthing, as detailed in
```
